src/ray/object_manager/plasma/test/object_lifecycle_manager_test.cc

```diff
@@ -0,0 +1,330 @@
+  EXPECT_EQ(expect_notified_ids, notify_deleted_ids_);
+}
+
+TEST_F(ObjectLifecycleManagerTest, AddReference) {
+  {
+    EXPECT_CALL(*object_store_, GetObject(id1_)).Times(1).WillOnce(Return(nullptr));
+    EXPECT_FALSE(manager_->AddReference(id1_));
+  }
+
+  {
+    EXPECT_CALL(*object_store_, GetObject(id2_)).Times(1).WillOnce(Return(&object1_));
+    EXPECT_CALL(*eviction_policy_, BeginObjectAccess(id2_)).Times(1).WillOnce(Return());
+    EXPECT_TRUE(manager_->AddReference(id2_));
+    EXPECT_EQ(1, object1_.GetRefCount());
+  }
+
+  {
+    EXPECT_CALL(*object_store_, GetObject(id3_))
+        .Times(1)
+        .WillOnce(Return(&one_ref_object_));
+    EXPECT_TRUE(manager_->AddReference(id3_));
+    EXPECT_EQ(2, one_ref_object_.GetRefCount());
+  }
+}
+
+TEST_F(ObjectLifecycleManagerTest, RemoveReferenceFailure) {
+  {
+    EXPECT_CALL(*object_store_, GetObject(id1_)).Times(1).WillOnce(Return(nullptr));
+    EXPECT_FALSE(manager_->RemoveReference(id1_));
+  }
+
+  {
+    EXPECT_CALL(*object_store_, GetObject(id2_)).Times(1).WillOnce(Return(&object1_));
+    EXPECT_FALSE(manager_->RemoveReference(id2_));
+  }
+}
+
+TEST_F(ObjectLifecycleManagerTest, RemoveReferenceTwoRef) {
+  EXPECT_CALL(*object_store_, GetObject(id1_))
+      .Times(1)
+      .WillOnce(Return(&two_ref_object_));
+  EXPECT_TRUE(manager_->RemoveReference(id1_));
+  EXPECT_EQ(1, two_ref_object_.GetRefCount());
+}
+
+TEST_F(ObjectLifecycleManagerTest, RemoveReferenceOneRefSealed) {
+  EXPECT_TRUE(one_ref_object_.Sealed());
+  EXPECT_CALL(*object_store_, GetObject(id1_))
+      .Times(1)
+      .WillOnce(Return(&one_ref_object_));
+  EXPECT_CALL(*eviction_policy_, EndObjectAccess(id1_)).Times(1).WillOnce(Return());
+  EXPECT_TRUE(manager_->RemoveReference(id1_));
+  EXPECT_EQ(0, one_ref_object_.GetRefCount());
+}
+
+TEST_F(ObjectLifecycleManagerTest, RemoveReferenceOneRefNotSealed) {
+  one_ref_object_.state = ObjectState::PLASMA_CREATED;
+  EXPECT_CALL(*object_store_, GetObject(id1_))
+      .Times(1)
+      .WillOnce(Return(&one_ref_object_));
+  EXPECT_CALL(*eviction_policy_, EndObjectAccess(id1_)).Times(1).WillOnce(Return());
+  EXPECT_TRUE(manager_->RemoveReference(id1_));
+  EXPECT_EQ(0, one_ref_object_.GetRefCount());
+}
+
+TEST_F(ObjectLifecycleManagerTest, RemoveReferenceOneRefEagerlyDeletion) {
+  manager_->eargerly_deletion_objects_.emplace(id1_);
+
+  EXPECT_CALL(*object_store_, GetObject(id1_))
+      .Times(2)
+      .WillRepeatedly(Return(&one_ref_object_));
+  EXPECT_CALL(*eviction_policy_, EndObjectAccess(id1_)).Times(1).WillOnce(Return());
+  EXPECT_CALL(*object_store_, DeleteObject(id1_)).Times(1).WillOnce(Return(true));
+  EXPECT_CALL(*eviction_policy_, RemoveObject(id1_)).Times(1).WillOnce(Return());
+
+  EXPECT_TRUE(manager_->RemoveReference(id1_));
+  EXPECT_EQ(0, one_ref_object_.GetRefCount());
+
+  std::vector<ObjectID> expect_notified_ids{id1_};
+  EXPECT_EQ(expect_notified_ids, notify_deleted_ids_);
+}
+}  // namespace plasma
+
```

Comment:
if you don't plan to add RAY_CHECK on add reference / remove reference, can you add tests to verify there's no issue when they are called multiple times on the boundary around 0? For example, Calling -> Remove Reference 3 times + Add reference 2 times -> should have ref == 2

Reply:
oh it's covered by RemoveReferenceFailure actually!